07-2d-parallelism/train_llm.py

```diff
@@ -0,0 +1,487 @@
+            # NOTE: not sure if we can do sequence parllel on this one? due to class construction
+            "model.norm": tp.SequenceParallel(),
+            "lm_head": tp.ColwiseParallel(
+                # Assuming we can do sequence parallel for model.norm, we will be receiving the sharded input
+                input_layouts=Shard(1),
+                # and we want the outputs to be sharded on the last dimension, which is the default behavior,
+                # but we need to convert this to a Tensor
+                use_local_output=False,
+            ),
+        },
+    )
+    for layer in model.model.layers:
+        # Have the adjust these values since we are sharding the linear layers
+        layer.self_attn.num_heads //= mesh["tp"].size()
+        layer.self_attn.num_key_value_heads //= mesh["tp"].size()
+
+        tp.parallelize_module(
+            layer,
+            mesh["tp"],
+            {
+                # SequenceParallel will apply sharding to sequence dimension.
+                "input_layernorm": tp.SequenceParallel(),
+                # The input to self_attn (which is the output from the SequenceParallel input_layer_norm) will be sharded on dimension 1, but we wanted it to be the whole tensor.
+                "self_attn": tp.PrepareModuleInput(
+                    input_layouts=Shard(dim=1), desired_input_layouts=Replicate()
+                ),
+                # TODO talk about how this all works together (go through line by line in the flash attn forward to show dims)
+                "self_attn.q_proj": tp.ColwiseParallel(),
+                "self_attn.k_proj": tp.ColwiseParallel(),
+                "self_attn.v_proj": tp.ColwiseParallel(),
+                "self_attn.o_proj": tp.RowwiseParallel(),
+                # TODO should we apply tensor parallel to self_attn?
+                # Another sharding along sequence dimension.
+                "post_attention_layernorm": tp.SequenceParallel(),
+                "mlp": tp.PrepareModuleInput(
+                    input_layouts=Shard(dim=1), desired_input_layouts=Replicate()
+                ),
+                # TODO Show graphic from pytorch lightning for explaining this.
+                "mlp.gate_proj": tp.ColwiseParallel(),
+                "mlp.up_proj": tp.ColwiseParallel(),
+                "mlp.down_proj": tp.RowwiseParallel(),
+            },
+        )
+
+    LOGGER.info(f"Before FSDP: {_get_cuda_mem_stats()}")
+    torch.cuda.empty_cache()
+
+    model = FullyShardedDataParallel(
+        model,
+        device_id=local_rank,
+        param_init_fn=lambda m: m.to_empty(device=device, recurse=False),
+        auto_wrap_policy=functools.partial(
+            transformer_auto_wrap_policy,
+            transformer_layer_cls={LlamaDecoderLayer},
+        ),
+        cpu_offload=CPUOffload(offload_params=True),
+        sharding_strategy=ShardingStrategy.NO_SHARD,
+        device_mesh=mesh["dp"],
```

Comment:
It seems like if you pass in sync_module_states=True, and HYBRID_SHARD, that it will broadcast from the right process group.

I'd test this out. But it also means we'd need to load the weights on ranks 0-7?